File: src/ELDCalculator.java

/**
 * Calculator class for performing Lambda Iteration-based Economic Load Dispatch.
 */
public class ELDCalculator {
    private float lambda;
    private Generator[] genArray;
    private int numGenerators;
    private float totDemand;
    private float tolerance;
    private int maxIterations;

    /**
     * Constructor to initialize ELDCalculator with generator data and demand.
     *
     * @param numGenerators number of generators
     * @param genArray array of Generator objects
     * @param totDemand total power demand
     */
    public ELDCalculator(int numGenerators, Generator[] genArray, float totDemand) {
        this.lambda = 1f;
        this.genArray = genArray;
        this.numGenerators = numGenerators;
        this.totDemand = totDemand;
        this.tolerance = 0.001f;
        this.maxIterations = 100;
    }

    /**
     * Performs Lambda Iteration to compute optimal power distribution.
     *
     * @return array of dispatched power values for each generator
     */
    public float[] lambdaIteration() {
        float[] P = new float[numGenerators];
        int iteration = 0;

        while (iteration < maxIterations) {
            float totalPower = 0f;

            for (int i = 0; i < numGenerators; i++) {
                float a = genArray[i].getA();
                float b = genArray[i].getB();
                float c = genArray[i].getC();

                P[i] = (lambda - b) / (2 * c);
                P[i] = genArray[i].validatePower(P[i]);
                totalPower += P[i];
            }

            if (Math.abs(totalPower - totDemand) <= tolerance) {
                System.out.println("✅ Converged Economic Load Dispatch values:");
                for (int i = 0; i < P.length; i++) {
                    System.out.printf("Generator %d : %.3f kW%n", i + 1, P[i]);
                }
                break;
            } else if (totalPower < totDemand) {
                lambda += 0.01f;
            } else {
                lambda -= 0.01f;
            }

            iteration++;
        }

        if (iteration == maxIterations) {
            System.out.println("⚠️ Max iterations reached. Solution may not be optimal.");
        }

        return P;
    }
}

File: src/main/python/eld_calculator.py
from generator import Generator


class ELDCalculator:

    def __init__(self, generators, total_demand):
        self.generators = generators
        self.total_demand = total_demand
        self.tolerance = 0.001
        self.max_iterations = 100
        self.lambda_value = 0
        self.lambda_history = []
        self.power_history = []

    def lambda_iteration(self):

        n = len(self.generators)
        power = [0.0] * n

        lambda_min = float("inf")
        lambda_max = float("-inf")

        for g in self.generators:

            min_lambda = g.b + 2 * g.c * g.min_cap
            max_lambda = g.b + 2 * g.c * g.max_cap

            lambda_min = min(lambda_min, min_lambda)
            lambda_max = max(lambda_max, max_lambda)

        iteration = 0

        while iteration < self.max_iterations:

            self.lambda_value = (lambda_min + lambda_max) / 2

            total_power = 0

            for i, g in enumerate(self.generators):

                power[i] = (self.lambda_value - g.b) / (2 * g.c)

                power[i] = g.validate_power(power[i])

                total_power += power[i]

            self.lambda_history.append(self.lambda_value)
            self.power_history.append(total_power)

            error = total_power - self.total_demand

            print(
                f"Iter {iteration:3d} | "
                f"Lambda={self.lambda_value:.5f} | "
                f"Power={total_power:.3f}"
            )

            if abs(error) <= self.tolerance:

                print("\n✅ Economic Load Dispatch Converged")
                print(f"Lambda = {self.lambda_value:.5f}")

                for i, g in enumerate(self.generators):
                    print(
                        f"eld_lambda.Generator {g.gen_id} Output = "
                        f"Power {power} MW"
                    )

                print(
                    f"Total Generated Power = "
                    f"{total_power:.3f} MW"
                )

                return power

            if total_power < self.total_demand:
                lambda_min = self.lambda_value
            else:
                lambda_max = self.lambda_value

            iteration += 1

        print("⚠️ Maximum iterations reached")
        return power